dhall-docs/README.md

```diff
@@ -97,6 +60,281 @@ dhall-docs --input . --output-link ${OTHER_LINK}
+               = \(x : Natural) -> x + 1
+
+in
+{
+  {-|
+  foo
+      bar
+          baz
+  -}
+  myAwesomeFunction
+}
+```
+
+If a block comment doesn't starts with `{-|` it will be ignored by `dhall-docs`.
+On the other hand, if a block comments starts with `{-|` and it doesn't follow
+the rules on the specification, it will be ignored _and_ a warning will be
+logged by `dhall-docs`.
+
+### Line comments
+
+Normal Dhall single-line comments starts with `--` and ends at a newline. For
+example:
+
+```dhall
+-- single line comment
+0
+```
+
+A `dhall-docs` valid single-line comment starts with `--|` and a single whitespace
+character. For example:
+
+```dhall
+--| foo
+```
+
+You can span your documentation in several line comments by writing `--` and two
+(2) whitespaces:
+
+```dhall
+--| foo
+--  bar
+--      baz
+```
+
+The `--` prefix defines the indentation for any line. That is, on the above
+example and the following
+
+```dhall
+--| foo
+  --  bar
+ --      baz
+```
+
+... `dhall-docs` will extract the following contents:
+
+```
+foo
+bar
+    baz
+```
+
+If you would like to force an empty line on the documentation output, place an
+empty single line comment in that place. On this example:
+
+```dhall
+--| foo
+--
+--  bar
+```
+
+`dhall-docs` will extract:
+
+```dhall
+foo
+
+bar
+```
+
+The empty `--` line doesn't need the two (2) whitespaces.
+
+If you left an empty line on the source code between two lines, it will be ignored.
+On this example:
+
+```dhall
+--| foo
+
+--  bar
+
+```
+
+`dhall-docs` will extract:
+
+```
+foo
+bar
+```
+
+Here is an example of using this type of comments in a Dhall file:
+
+```dhall
+--| # My header
+--
+--  * 1
+--  * 2
+
+let a --| foo
+        --  bar
+        = True
+
+in { a }
+```
+
+Although `dhall-docs` doesn't enfoce an alignment between the set of consecutive
+single-line comments, we encourage you to write them aligned. The above example
+written with this recommendation looks like this:
+
+```dhall
+--| # My header
+--
+--  * 1
+--  * 2
+
+let a --| foo
+      --  bar
+        = True
+
+in { a }
+
+```
+
+To end, if the first line of a single-line comments doesn't start with `--|` it
+will be ignored by `dhall-docs`. On the other hand if the first line starts with
+`--|` but the set of lines doesn't follow the rules on this specification, it
+will be ignored _and_ a warning will be logged to the console.
+
+### Supported annotated elements
+
+A limitation of the `dhall` parser is that it doesn't preserve _all_ the whitespaces
+of a Dhall source file, but preserves the enough to write useful documentation.
+`dhall-docs` supports the following comments:
+
+* The header of a file:
+
+    ```dhall
+    {-| this is the header
+    -}
+    let a = 1
+    in  a
+    ```
+
+* Before each key-value pair of a record type and literal:
+
+    ```dhall
+    {
+      --| here
+      a : T,
+    }
+
+    {
+      -- | and here
+      pun,
+
+      -- here too
+      b = True
+    }
+    ```
```

Comment:
I think we should maybe wait with this until we've worked out what `dhall-docs` will do with these annotations.